cgroup: devices: major cleanups and minimal transition rules

cyphar opened the pull request:
#2366 is mostly fixed by this, though the patch is unfortunately a little more complicated than you might've first expected (it includes an emulator for the devices controller). Unfortunately this logic is necessary in order to be able to calculate and verify the minimal transition rules for a cgroup update.

A similar issue affects the cgroupv2 devices setup, but that is a topic for another time (as the solution is drastically different).

Fixes #2204
Signed-off-by: Aleksa Sarai <asarai@suse.de>

cyphar commented:
@mYmNeo does this PR fix the original issue for you?

cyphar commented:
According to my testing, this does fix #2204.

kolyshkin commented on libcontainer/cgroups/fs/devices.go:
```diff
@@ -31,44 +31,24 @@ func (s *DevicesGroup) Set(path string, cgroup *configs.Cgroup) error {
-			}
-		}
-		return nil
+	// The devices list is a blacklist, so we must first deny all devices.
```
do you mean `whitelist` here?

cyphar replied:
Yup, will fix.

kolyshkin commented on libcontainer/cgroups/fs2/devices.go:
```diff
@@ -3,35 +3,27 @@
+func isRWM(perm string) bool {
+	r := strings.Contains(perm, "r")
+	w := strings.Contains(perm, "w")
+	m := strings.Contains(perm, "m")
```
Looks like the old implementation is about 3x more efficient. Not that this is a critical part :smile: but I don't see a reason to replace it.

kolyshkin replied:
I would do that for readability but to my mind both the old and the new one is fine.

cyphar replied:
Yeah, I'm not actually sure why I did this tbh. :thinking:

kolyshkin commented on libcontainer/configs/device.go:
```diff
@@ -35,23 +28,153 @@ type Device struct {
+	}
+	if strings.Contains(string(p), "m") {
+		set |= deviceMknod
+	}
```
1. Is it OK to ignore unknown characters here?

2. I would rewrite it to do
```go
for _, c := range string(p) {
  switch c {
  case 'r':
    set |= deviceRead
  case 'w':
......
....
```

kolyshkin replied:
> Is it OK to ignore unknown characters here?

oh, I see you have `IsValid()` below, nice.

cyphar replied:
Yeah, I'll switch this to also use the iterator setup.

AkihiroSuda commented:
Can we have integration tests?

kolyshkin commented on libcontainer/configs/device.go:
```diff
@@ -35,23 +28,153 @@ type Device struct {
-		return "*"
+func (d *DeviceRule) Mkdev() (uint64, error) {
+	if d.Major == Wildcard || d.Minor == Wildcard {
+		return 0, fmt.Errorf("cannot mkdev() device with wildcards")
```
nit: no need to use formatting here since there's no arguments to format, so `errors.New()` is better.

cyphar replied:
Sure, though I personally don't like mixing `fmt` and `errors` in the same file.

kolyshkin commented on libcontainer/configs/device.go:
```diff
@@ -35,23 +28,153 @@ type Device struct {
+func (d *DeviceRule) CgroupString() string {
+	var (
+		major = fmt.Sprint(d.Major)
+		minor = fmt.Sprint(d.Minor)
```
nit: better use `strconf.Itoa()` which doesn't have to reflect on the argument type and is faster.

kolyshkin commented on libcontainer/cgroups/fs/devices.go:
```diff
@@ -26,29 +31,74 @@ func (s *DevicesGroup) Apply(d *cgroupData) error {
+	//
+	// This safety-check is skipped for the unit tests because we cannot
+	// currently mock devices.list correctly.
+	if !s.testingSkipFinalCheck {
```
nit:
```go
if s.testingSkipFinalCheck {
  return nil
}
....
```

cyphar replied:
I disagree, I think that an early-return like that makes it harder to read.

kolyshkin replied:
To me earlier return is almost always more readable, but only marginally in this case so whatever works for you

mYmNeo commented:
@cyphar I've tested these patches, it works well. Thank you very much. 👍

cyphar commented:
@AkihiroSuda Working on it.

cyphar commented:
@crosbymichael I'd like to know if you're okay with the `/dev/console` rule removal as well as some of the core refactors. Quite a bit of this code predates even my involvement in `libcontainer`.

cyphar commented:
@AkihiroSuda Integration test added.

AkihiroSuda commented on tests/integration/update.bats:
```diff
@@ -312,3 +312,42 @@ EOF
+	CONTAINER_OUTPUT="$TMP_RECVTTY_DIR/output"
+	("$RECVTTY" --no-stdin --pid-file "$TMP_RECVTTY_PID" \
+				--mode single "$TMP_CONSOLE_SOCKET" &>"$CONTAINER_OUTPUT" &) &
+	sleep 0.5s
```
maybe too short, `sleep 1s` ?

cyphar replied:
I didn't have issues with my testing, but sure I'll bump it up.

AkihiroSuda commented on tests/integration/update.bats:
```diff
@@ -312,3 +312,42 @@ EOF
+	done
+
+	# Kill recvtty.
+	kill -9 "$(<"$TMP_RECVTTY_PID")"
```
Is SIGKILL needed?

cyphar replied:
We do `kill -9` for the regular `recvtty` setup, I just copied it.

AkihiroSuda commented on libcontainer/cgroups/devices_emulator.go:
```diff
@@ -0,0 +1,351 @@
+// +build linux
+
+package cgroups
```
Maybe we want to create a dedicated package for `libcontainer/cgroups/devicesemulator`

cyphar replied:
I've put it in `libcontainer/cgroups/devices` so we can write `devices.Emulator`.

AkihiroSuda commented on libcontainer/cgroups/devices_emulator.go:
```diff
@@ -0,0 +1,351 @@
+
+func (e *DevicesEmulator) deny(rule *deviceRule) error {
+	// This cgroup is configured as a white-list. Reset the entire emulator,
+	// and put is into white-list mode.
```
put what?

cyphar replied:
Typo, should be "put it".

AkihiroSuda commented on libcontainer/cgroups/ebpf/devicefilter/devicefilter_test.go:
```diff
@@ -83,71 +83,69 @@ block-2:
         20: JNEImm dst: r2 off: -1 imm: 2 <block-4>
-        21: JNEImm dst: r4 off: -1 imm: 5 <block-4>
-        22: JNEImm dst: r5 off: -1 imm: 1 <block-4>
+        21: JNEImm dst: r4 off: -1 imm: 1 <block-4>
```
How are these changes introduced?

cyphar replied:
The removal of the `/dev/console` rule means that `5:1` is no longer in the default whitelist, which changes the code generated.

AkihiroSuda commented on libcontainer/configs/device.go:
```diff
@@ -35,23 +29,154 @@ type Device struct {
+// IsEmpty returns whether the set of permissions in a DevicePermissions is
+// empty.
+func (p DevicePermissions) IsEmpty() bool {
+	return p == DevicePermissions("")
```
Maybe needs `strings.TrimSpace` somewhere

cyphar replied:
We could add it, but there's no point where we parse `DevicePermissions` that would allow a space to be added -- there's `IsValid` checks everywhere relevant. Honestly this whole `DevicePermissions` thing is to work around the fact that we (stupidly) embedded the string representation into the libcontainer state format...

cyphar commented:
The new integration test fails under the systemd cgroup driver. I'll need to figure out what's going on there.

kolyshkin commented on tests/integration/update.bats:
```diff
@@ -312,3 +312,46 @@ EOF
+	TMP_CONSOLE_SOCKET="$TMP_RECVTTY_DIR/console.sock"
+	CONTAINER_OUTPUT="$TMP_RECVTTY_DIR/output"
+	("$RECVTTY" --no-stdin --pid-file "$TMP_RECVTTY_PID" \
+				--mode single "$TMP_CONSOLE_SOCKET" &>"$CONTAINER_OUTPUT" &) &
```
I know this comes from `helpers.bats` but the double `&` is not really needed here, you'll be fine with something like
```shell
("$RECVTTY" ... ) &
```

IOW inner `&` is not needed.

kolyshkin commented on tests/integration/update.bats:
```diff
@@ -312,3 +312,46 @@ EOF
+	CONTAINER_OUTPUT="$TMP_RECVTTY_DIR/output"
+	("$RECVTTY" --no-stdin --pid-file "$TMP_RECVTTY_PID" \
+				--mode single "$TMP_CONSOLE_SOCKET" &>"$CONTAINER_OUTPUT" &) &
+	sleep 1s
```
I hate sleeps in the test code :( it always means we're not doing something right

In this case, I guess, there is no way to check that recvtty is ready.

Or can we maybe wait till "$TMP_CONSOLE_SOCKET" is created? Something like
```shell
retry 10 0.1 test -e "$TMP_CONSOLE_SOCKET"
```
?

cyphar replied:
Yeah that'd work.

kolyshkin commented on tests/integration/update.bats:
```diff
@@ -312,3 +312,46 @@ EOF
+
+	# Trigger an update. This update doesn't actually change the device rules,
+	# but it will trigger the devices cgroup code to reapply the current rules.
+	# We trigger the update a few times to make sure we hit the race.
```
Slow Saturday here... which race?

cyphar replied:
That the container will try to write to `/dev/null` during the window between `a *:* rwm` being denied and the rest of the white-list being applied.

AkihiroSuda replied:
Please add code comments for the race?

cyphar commented:
Ah, I think I know what the issue is. We start a transient unit under systemd even when we do `runc update` -- so systemd clears all of the state. We'll need to do something different if the container is already started (does systemd support an "update" to a transient unit).

This is a little scary because during the race window, systemd actually switches the devices cgroup **to be allow-all** (since that's the default). Oops.

cyphar commented:
Ah, I think I know what the issue is. We start a transient unit under systemd even when we do `runc update` -- so systemd clears all of the state. We'll need to do something different if the container is already started. In particular, we need to use `SetUnitProperties` if the container is already running.

This is quite scary because during the race window, systemd actually switches the devices cgroup **to be allow-all** (since that's the default). Which means that a container could theoretically get full access to any block or character device on the system. Oops. Yet another reason to not use the systemd cgroup manager I guess :wink:.

cyphar commented:
I just implemented the above, but it turns out that systemd has the same bug outlined above (if you tell it to reset `DevicesAllow` and then give it a new allow list, it will always write a deny-all blacklist first). So we'll just need to disable this test for systemd cgroups, since this behaviour is coming from systemd.

cyphar commented:
Since this PR solves a few actual security bugs, I might request a CVE. The systemd cgroup one is probably worth a CVE on its own because it means containers could do all sorts of worrying attacks...

cyphar commented:
Ah, of course. `DevicesAllow` doesn't allow wildcards for `/dev/{char,block}`. This is going to be quite ugly...

AkihiroSuda commented:
Can we use freezer to mitigate the systemd race?

cyphar commented:
"I used the cgroups to manage the cgroups."

Yeah, that's not a bad idea. Working on it.

cyphar commented:
Hurray, all green! :green_heart:

AkihiroSuda commented on libcontainer/cgroups/systemd/v1.go:
```diff
@@ -394,6 +418,13 @@ func (m *legacyManager) Set(container *configs.Config) error {
 		return err
 	}
 
+	// Unfreeze the cgroup.
```
Seems covered in `defer m.Freeze(configs.Thawed)`

cyphar replied:
I wanted to make sure we unfroze the cgroup before we ran the fsManager (to limit the length the time the container remains frozen, as well as to avoid stepping on our toes with the fsManager also writing to the freezer cgroup).

However, thinking about this more the `defer` is incorrect (as is using `configs.Thawed` directly) because if you're dealing with a user-frozen container (or the user requested we create a frozen container) then we won't do what they're expecting. I'll fix that now.

AkihiroSuda commented on tests/integration/update.bats:
```diff
@@ -312,3 +312,66 @@ EOF
+@test "update devices [minimal transition rules]" {
+    [[ "$ROOTLESS" -ne 0 ]] && requires rootless_cgroup
+	# This test currently only makes sense on cgroupv1.
+    requires cgroups_v1
```
indentation is broken

(off-topic: we should have `shfmt` in CI)

AkihiroSuda commented on tests/integration/update.bats:
```diff
@@ -312,3 +312,66 @@ EOF
+}
+
+@test "update paused container" {
+    [[ "$ROOTLESS" -ne 0 ]] && requires rootless_cgroup
```
indentation

AkihiroSuda commented:
LGTM

shfmt issue can be handled in follow-up PR

cyphar commented:
Hang on, I'm still working on fixing the tests -- the newest one is breaking (the auto-detection for the freezer cgroup isn't working). Should only take a moment to fix...

cyphar commented:
Alright, this is ready for review. Tests are implemented for the changes, and I enabled the `runc pause` tests for rootless containers which was missed some time ago.

kolyshkin commented:
LGTM

cyphar commented:
Rebased, now that the `GetPaths` changes have been merged.

/cc @opencontainers/runc-maintainers

AkihiroSuda commented:
LGTM

AkihiroSuda commented on tests/integration/helpers.bash:
```diff
@@ -200,75 +207,85 @@ function fail() {
 			;;
 		esac
+		if [ -n "$skip" ]; then
+			skip "test requires $var"
```
nit: Can we avoid using the same identifier for the variable and the function. (Surprised to know they don't conflict)

cyphar replied:
Sure. The reason they don't conflict is that variables and functions are in different namespaces.

cyphar replied:
Done.

AkihiroSuda commented on tests/integration/helpers.bash:
```diff
@@ -200,75 +207,85 @@ function fail() {
 # support it, the test is skipped with a message.
 function requires() {
 	for var in "$@"; do
+		SKIP_ME=
```
nit again: `local skip_me` (in lower characters)

AkihiroSuda commented:
LGTM

cyphar commented:
/cc @kolyshkin

cyphar commented:
I've rebased and cleaned up the commit message for the `DeviceAllow` commit to better clarify that it was a security issue but since no released version of runc included it, it didn't deserve a proper security advisory.

kolyshkin commented:
LGTM

AkihiroSuda commented:
LGTM

tao12345666333 commented:
When I try to vendor runc library v1.0.0-rc91 into Docker/Moby,  I found an unexpected behavior: https://github.com/moby/moby/pull/41178#issuecomment-663813218

After debugging, I found that the problem was introduced since this PR.  PTAL, Thanks!  https://github.com/moby/moby/pull/41178

AkihiroSuda commented:
Could you open a new issue?

tao12345666333 commented:
sure.
